drivers/mdio/mdio_stm32_hal.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024 BayLibre, SAS
+ * Copyright (c) 2010-2024 Analog Devices Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <errno.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/mdio.h>
+#include <zephyr/net/ethernet.h>
+#include <zephyr/net/mdio.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mdio_stm32_hal, CONFIG_MDIO_LOG_LEVEL);
+
+#define DT_DRV_COMPAT st_stm32_mdio
+
+struct mdio_stm32_data {
+	struct k_sem sem;
+	ETH_HandleTypeDef hmdio;
+};
+
+struct mdio_stm32_config {
+	const struct pinctrl_dev_config *pincfg;
+};
+
+static int mdio_stm32_read(const struct device *dev, uint8_t prtad,
+			   uint8_t regad, uint16_t *data)
+{
+	struct mdio_stm32_data *const dev_data = dev->data;
+	ETH_HandleTypeDef *hmdio = &dev_data->hmdio;
+	uint32_t read;
+	int ret;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+
+	ret = HAL_ETH_ReadPHYRegister(hmdio, prtad, regad, &read);
+
+	k_sem_give(&dev_data->sem);
+
+	if (ret != HAL_OK)
+		return -EIO;
+
+	*data = read & 0xffff;
+
+	return ret;
+}
+
+static int mdio_stm32_write(const struct device *dev, uint8_t prtad,
+			    uint8_t regad, uint16_t data)
+{
+	struct mdio_stm32_data *const dev_data = dev->data;
+	ETH_HandleTypeDef *hmdio = &dev_data->hmdio;
+	int ret;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+
+	ret = HAL_ETH_WritePHYRegister(hmdio, prtad, regad, data);
```

Comment:
I noticed that the stm32 hal looks for definition of  ETH_MDIO_BUS_TIMEOUT to decide how long to block and I am wondering if it would be worth any value to provide this definition based on the value of a kconfig so that it is configurable

Reply:
It is,

#define ETH_MDIO_BUS_TIMEOUT    1000U

So it's actually 1 second max. It's a max limit, so busy flag is cleared to 0 much before.  Seems not simple for me  to replace this value, since part of stm32cube module repo.

Reply:
you could make a kconfig and give zephyr_compile_definitions in the cmake based on that config, but if you don't think this is worth it then I don't think it is a big deal to skip it in this PR